src/libraries/CalldataDecoder.sol

```diff
@@ -11,35 +11,61 @@ library CalldataDecoder {
 
     error SliceOutOfBounds();
 
-    /// @notice equivalent to SliceOutOfBounds.selector
-    bytes4 constant SLICE_ERROR_SELECTOR = 0x3b99b53d;
+    /// @notice mask used for offsets and lengths to ensure no overflow
+    /// @dev no sane abi encoding will pass in an offset or length greater than type(uint32).max
+    ///      (note that this does deviate from standard solidity behavior and offsets/lengths will
+    ///      be interpreted as mod type(uint32).max which will only impact malicious/buggy callers)
+    uint256 constant OFFSET_OR_LENGTH_MASK = 0xffffffff;
+    uint256 constant OFFSET_OR_LENGTH_MASK_AND_WORD_ALIGN = 0xffffffe0;
 
-    /// @dev equivalent to: abi.decode(params, (bytes, bytes[])) in calldata
+    /// @notice equivalent to SliceOutOfBounds.selector, stored in least-significant bits
+    uint256 constant SLICE_ERROR_SELECTOR = 0x3b99b53d;
+
+    /// @dev equivalent to: abi.decode(params, (bytes, bytes[])) in calldata (requires strict abi encoding)
```

Comment:
Might be better to note here that the masking makes it not quite strictly equivalent.